File: src/main/java/me/vaan/movecraft/expansions/MovecraftExpansion.java
package me.vaan.movecraft.expansions;

import me.clip.placeholderapi.expansion.PlaceholderExpansion;
import net.countercraft.movecraft.craft.Craft;
import net.countercraft.movecraft.craft.CraftManager;
import net.countercraft.movecraft.craft.PlayerCraft;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class MovecraftExpansion extends PlaceholderExpansion {
    @NotNull
    public String getAuthor() {
        return "Vaan1310";
    }

    @NotNull
    public String getVersion() {
        return "1.0.4";
    }

    @NotNull
    public String getRequiredPlugin() {
        return "Movecraft";
    }

    public boolean canRegister() {
        return Bukkit.getPluginManager().getPlugin(this.getRequiredPlugin()) != null;
    }

    @NotNull
    public String getIdentifier() {
        return "movecraft";
    }

    @Nullable
    public String onPlaceholderRequest(@Nullable Player executor, @NotNull String params) {
        if (executor == null) {
            return null;
        }

        CraftManager manager = CraftManager.getInstance();
        if (manager == null) {
            return null;
        }

        PlayerCraft craft = Utils.craftOf(executor);
        if (craft == null) {
            return "Not on craft";
        }

        switch (params) {
            case "burning_fuel" -> {
                return String.valueOf(craft.getBurningFuel());
            }
            case "craft_speed" -> {
                if (!craft.getCruising()) {
                    return "0";
                }

                return String.valueOf(craft.getSpeed());
            }
            case "flyblocks" -> {
                var flyBlocks = craft.getDataTag(Craft.FLYBLOCKS);
                int sumFlyblocks = Utils.sumOf(flyBlocks.getKeySet(), flyBlocks::get);
                return String.valueOf(sumFlyblocks);
            }
            case "craft_name" -> {
                return craft.getName();
            }
            case "moveblocks" -> {
                var moveblocks = craft.getDataTag(Craft.MOVEBLOCKS);
                int sumMoveblocks = Utils.sumOf(moveblocks.getKeySet(), moveblocks::get);
                return String.valueOf(sumMoveblocks);
            }
            case "cruising_direction" -> {
                return craft.getCruiseDirection().toString();
            }
            case "is_player_pilot" -> {
                return craft.getPilot().equals(executor) ? "True" : "False";
            }
            case "is_player_cruising" -> {
                return craft.getCruising() ? "True" : "False";
            }
            case "craft_current_gear" -> {
                return String.valueOf(craft.getCurrentGear());
            }
            case "blocks_on_craft" -> {
                var materials = craft.getDataTag(Craft.MATERIALS);
                int sumMaterials = Utils.sumOf(materials.getKeySet(), materials::get);
                return String.valueOf(sumMaterials);
            }
        }

        return null;
    }
}

File: src/main/kotlin/me/vaan/movecraft/expansions/MovecraftExpansion.kt
package me.vaan.movecraft.expansions

import me.clip.placeholderapi.expansion.PlaceholderExpansion
import net.countercraft.movecraft.craft.Craft
import net.countercraft.movecraft.craft.CraftManager
import net.countercraft.movecraft.craft.type.RequiredBlockEntry
import net.countercraft.movecraft.util.Counter
import org.bukkit.Bukkit
import org.bukkit.entity.Player

class MovecraftExpansion : PlaceholderExpansion() {

    override fun getAuthor(): String {
        return "Vaan1310"
    }

    override fun getVersion(): String {
        return "1.0.3"
    }

    override fun getRequiredPlugin(): String {
        return "Movecraft"
    }

    override fun canRegister(): Boolean {
        return Bukkit.getPluginManager().getPlugin(requiredPlugin) != null
    }

    override fun getIdentifier(): String {
        return "movecraft"
    }

    override fun onPlaceholderRequest(executor: Player?, params: String): String? {
        executor ?: return null

        val manager = CraftManager.getInstance()
        manager ?: return null

        val craft = executor.craft
        craft ?: return "Not on craft"

        when(params) {
            "is_player_pilot" -> {
                return if (craft.pilot == executor) "True" else "False"
            }

            "moveblocks" -> {
                val counter: Counter<RequiredBlockEntry> = craft.getDataTag(Craft.MOVEBLOCKS)
                val sum = counter.keySet.sumOf(counter::get)
                return sum.toString()
            }

            "flyblocks" -> {
                val counter: Counter<RequiredBlockEntry> = craft.getDataTag(Craft.FLYBLOCKS)
                val sum = counter.keySet.sumOf(counter::get)
                return sum.toString()
            }

            "blocks_on_craft" ->  {
                val counter = craft.getDataTag(Craft.MATERIALS)
                var total = 0

                for (material in counter.keySet) {
                    total += counter.get(material)
                }

                return "$total"
            }

            "is_player_cruising" -> return if (craft.cruising) "True" else "False"
            "craft_name" -> return craft.name
            "craft_speed" -> {
                if (!craft.cruising) {
                    return "0"
                }

                return craft.speed.toString()
            }

            "craft_current_gear" -> return craft.currentGear.toString()
            "cruising_direction" -> return craft.cruiseDirection.toString()
            "burning_fuel" -> return craft.burningFuel.toString()
        }

        return null
    }
}
